beamer/state_machine.py

```diff
@@ -45,29 +46,32 @@ class Context:
     config: Config
 
 
-def process_event(event: Event, context: Context) -> bool:
+HandlerResult = Tuple[bool, Optional[List[Event]]]
```

Comment:
Regarding returning a list of events during process events, I wonder if this is the right way to go. This means that the state output of the state in the middle will not be processed.
This might work for the L1 resolution case (although I have not thought it through) but could be problematic for others in the future.

Have you thought about generating new events in `process_requests()` ?

Reply:
Another option that comes to mind is to just queue the returned events for the next run of the event loop, so that the requests and claim state machines would always be run before. What do you think about this?

Reply:
> I'm not quite sure whether this has been changed in the meantime, but previously the order and nature of events was irrelevant by design, i.e. we iterated multiple times over events until we could no longer advance. Has this been changed?

I don't think this changed, I just added a way for handlers to return new events.

Reply:
It's not about the order. As a new feature, the state machine itself can create new events which is looped back into the current list of events. As a consequence, those events might alternate the state of  a request in one go. 
For events coming from the blockchain this is totally fine, because we don't want to process a request on an old state. I'm just wondering if we might want to process a request based on an event coming from the chain, and as an output generating an event which again will be looped back into the event processor. 

Can you see the difference in the approaches?

Reply:
Yes, that why I proposed this:

> Another option that comes to mind is to just queue the returned events for the next run of the event loop, so that the requests and claim state machines would always be run before. What do you think about this?

Reply:
currently it `[E2, E3, E4]`, but another proposal above is to not process E4 in that round of events.

Reply:
As you described @istankovic order should not matter. The design still holds. But What Paul is suggesting is to not even put it in the current list, rather than postponing its processing to the next `cycle` (Maybe the term iteration is misleading here)

Terminology: 

iteration = one iteration over the list of events per cycle
cycle = every execution of `process_events()`

Reply:
I guess I was confused by this comment:

> Another option that comes to mind is to just queue the returned events for the next run of the event loop, so that the requests and claim state machines would always be run before.

My interpretation was that the ordering would matter, for some reason.
So if my new understating is correct, the only difference is, if we postpone the newly generated event for the next _cycle_, it would delay processing of that event for up to 1 second?
Thay may or may not be fine, depending on the nature of the generated event (can that event cause new events to be generated? etc.).

Reply:
I think your new understanding is correct. The main difference here is that `process_requests()` and `process_claims()` runs on the "intermediary" state of the object once, and then in the next cycle the new event is applied.

We may or may not want that. It was just a concers that we kind of leapfrog a certain state if we directly apply the new event.

Reply:
I see Freds point and implemented the system so that newly created events will be processed in the next cycle.

Reply:
> Also, while I'm here: do we need a list (mutability) here?

New events will usually be attached to some list inside the handler, so I don't think it would be worthwhile to convert them to a tuple in the end.